app/src/main/java/com/arctouch/codechallenge/home/data/repository/MoviesRepository.kt

```diff
@@ -0,0 +1,37 @@
+package com.arctouch.codechallenge.home.data.repository
+
+import com.arctouch.codechallenge.core.api.TmdbApi
+import com.arctouch.codechallenge.core.domain.model.Either
+import com.arctouch.codechallenge.core.domain.model.Failure
+import com.arctouch.codechallenge.core.domain.model.Movie
+import com.arctouch.codechallenge.home.data.cache.MoviesCache
+import java.net.UnknownHostException
+
+class MoviesRepository(private val api: TmdbApi) {
+
+    suspend fun getUpcomingMoviesFromApi(page: Long): Either<Failure, List<Movie>> {
+        return try {
+            val result = api.upcomingMoviesAsync(page = page).await()
+            val resultBody = result.body()
+            when {
+                result.isSuccessful && resultBody?.results?.isNotEmpty() == true -> Either.Right(resultBody.results)
+                result.isSuccessful && resultBody?.results?.isEmpty() == true -> Either.Left(Failure.NoDataAvailable())
+                else -> Either.Left(Failure.RequestError())
+            }
+        } catch (ex: UnknownHostException) {
+            ex.printStackTrace()
+            Either.Left(Failure.NetworkConnection())
+        } catch (ex: Exception) {
+            ex.printStackTrace()
+            Either.Left(Failure.RequestError())
+        }
+    }
+
+    fun getUpcomingMoviesFromCache(): List<Movie> {
```

Comment:
Usually the client should not know about the sources of the data, but in this case i have some situations where i need to get the data only from cache or api, so i needed two distinct methods, also because i based my code on Clean Architecture principles, and thoose principles say that i should keep all the control/bussiness logic on the usecases layer, all other layers must be clear and simple, if decide to load the data only from cache or only from api i don´t need to change the repository only my usecase rules, i could also reuse this repository for other app or screen where the rules are different, just writing another usecase